src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -0,0 +1,317 @@
+                    // stored in the table already.  If they don't match, don't read
+                    // out the data value at all.
+                    if (t.checksumOpt != null &&
+                        !t.self.ChecksumsMatch_MustRunInTransaction(t.connection, t.database, t.rowId, t.checksumOpt, t.cancellationToken))
+                    {
+                        return null;
+                    }
+
+                    return t.connection.ReadBlob_MustRunInTransaction(t.database, t.self.DataTableName, t.columnName, t.rowId);
+                };
+
+            private bool ChecksumsMatch_MustRunInTransaction(
+                SqlConnection connection, Database database, long rowId, Checksum checksum, CancellationToken cancellationToken)
+            {
+                using var checksumStream = connection.ReadBlob_MustRunInTransaction(database, DataTableName, ChecksumColumnName, rowId);
+                using var reader = ObjectReader.TryGetReader(checksumStream, leaveOpen: false, cancellationToken);
+
+                return reader != null && Checksum.ReadFrom(reader) == checksum;
+            }
+
+            protected bool GetAndVerifyRowId(SqlConnection connection, Database database, long dataId, out long rowId)
+            {
+                // For the Document and Project tables, our dataId is our rowId:
+                // 
+                // https://sqlite.org/lang_createtable.html
+                // if a rowid table has a primary key that consists of a single column and the 
+                // declared type of that column is "INTEGER" in any mixture of upper and lower 
+                // case, then the column becomes an alias for the rowid. Such a column is usually
+                // referred to as an "integer primary key". A PRIMARY KEY column only becomes an
+                // integer primary key if the declared type name is exactly "INTEGER"
+#if DEBUG
+                // make sure that if we actually request the rowId from the database that it
+                // is equal to our data id.  Only do this in debug as this can be expensive
+                // and we definitely do not want to do this in release.
+                if (TryGetRowIdWorker(connection, database, (TDatabaseId)(object)dataId, out rowId))
+                {
+                    Debug.Assert(dataId == rowId);
+                }
+#endif
+
+                // Can just return out dataId as the rowId without actually having to hit the 
+                // database at all.
+                rowId = dataId;
+                return true;
+            }
+
+            protected virtual bool TryGetRowId(SqlConnection connection, Database database, TDatabaseId dataId, out long rowId)
+                => TryGetRowIdWorker(connection, database, dataId, out rowId);
+
+            private bool TryGetRowIdWorker(SqlConnection connection, Database database, TDatabaseId dataId, out long rowId)
```

Comment:
Why the Worker method?

Reply:
so the top is virtual, with overrides.  but overall i don't like this.  i'll make abstract and try to make this clearer.

Reply:
ok.  hopefully clearer now.